File: app/src/main/java/com/faendir/lightning_launcher/multitool/scriptmanager/ScriptManagerFragment.java
package com.faendir.lightning_launcher.multitool.scriptmanager;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.Toast;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import com.faendir.lightning_launcher.multitool.R;
import com.faendir.lightning_launcher.multitool.event.ClickEvent;
import com.faendir.lightning_launcher.multitool.fastadapter.Model;
import com.faendir.lightning_launcher.multitool.util.Utils;
import com.faendir.lightning_launcher.scriptlib.ScriptManager;
import java9.util.stream.StreamSupport;
import org.acra.ACRA;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.faendir.lightning_launcher.multitool.MultiTool.DEBUG;

/**
 * Created by Lukas on 22.08.2015.
 * Main activity of ScriptManager
 */
public class ScriptManagerFragment extends Fragment implements ActionMode.Callback {
    private static final int IMPORT = 1;
    private static final int EXPORT = 2;
    private SharedPreferences sharedPref;
    private boolean enableMenu;
    private FrameLayout layout;
    private ListManager listManager;
    private ScriptManager scriptManager;
    private ActionMode actionMode;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        scriptManager = new ScriptManager(getActivity());
        if (DEBUG) scriptManager.enableDebug();
        sharedPref = PreferenceManager.getDefaultSharedPreferences(getActivity());
        listManager = new ListManager(scriptManager, getActivity(), this::setActionModeEnabled);
        setHasOptionsMenu(true);
        enableMenu = false;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        layout = new FrameLayout(getActivity());
        return layout;
    }

    @Override
    public void onStart() {
        super.onStart();
        EventBus.getDefault().register(this);
        loadFromLauncher();
    }

    @Override
    public void onStop() {
        EventBus.getDefault().unregister(this);
        super.onStop();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_scriptmanager, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (!enableMenu) {
            Toast.makeText(getActivity(), R.string.toast_menuDisabled, Toast.LENGTH_SHORT).show();
            return true;
        }
        switch (item.getItemId()) {
            case R.id.action_restore:
                Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
                intent.addCategory(Intent.CATEGORY_OPENABLE);
                intent.setType("*/*");
                startActivityForResult(intent, IMPORT);
                break;
            case R.id.action_search:
                ScriptUtils.searchDialog(getActivity(), listManager);
                break;
            default:
                return super.onOptionsItemSelected(item);
        }
        return true;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            switch (requestCode) {
                case IMPORT:
                    if (data != null && data.getData() != null) {
                        ScriptUtils.restoreFromFile(scriptManager, getActivity(), listManager, data.getData());
                    }
                    break;
                case EXPORT:
                    if (data != null && data.getData() != null) {
                        ScriptUtils.backup(getActivity(), listManager, (Script) listManager.getSelectedItems().get(0), data.getData());
                    }
            }
        }
    }

    private void loadFromLauncher() {
        scriptManager.getAsyncExecutorService().setKeepAliveAfterwards(true).add(ScriptUtils.getScriptManagerExecutor(null), this::handleScriptResult).start();
        layout.removeAllViews();
        LayoutInflater.from(getActivity()).inflate(R.layout.fragment_loading, layout);
    }

    private void handleScriptResult(@Nullable String result) {
        if (result != null) {
            List<Script> scripts = Arrays.asList(Utils.GSON.fromJson(result, Script[].class));
            listManager.updateFrom(scripts);
            listManager.setAsContentOf(layout);
            enableMenu = true;
        }
    }

    @Subscribe
    public void onReloadButton(ClickEvent event) {
        if (event.getId() == R.id.button_reload) {
            sharedPref.edit().putInt(getString(R.string.pref_id), -1).apply();
            loadFromLauncher();
        }
    }

    private void setActionModeEnabled(boolean enable) {
        if (enable) {
            if (actionMode == null) {
                actionMode = getActivity().startActionMode(this);
            } else {
                actionMode.invalidate();
            }
        } else if (actionMode != null) {
            actionMode.finish();
            actionMode = null;
        }
    }

    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
        mode.getMenuInflater().inflate(R.menu.menu_context_scriptmanager, menu);
        onPrepareActionMode(mode, menu);
        return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
        int selectionMode = listManager.getSelectionMode();
        if (selectionMode == ListManager.NONE) mode.finish();
        final boolean isOneScript = selectionMode == ListManager.ONE_SCRIPT;
        final boolean onlyScripts = isOneScript || selectionMode == ListManager.ONLY_SCRIPTS;
        menu.findItem(R.id.action_rename).setVisible(selectionMode == ListManager.ONE_GROUP || isOneScript);
        menu.findItem(R.id.action_edit).setVisible(isOneScript);
        menu.findItem(R.id.action_backup).setVisible(isOneScript);
        menu.findItem(R.id.action_format).setVisible(onlyScripts);
        MenuItem disable = menu.findItem(R.id.action_disable).setVisible(isOneScript);
        if (isOneScript) {
            disable.setTitle(((Script) StreamSupport.stream(listManager.getSelectedItems()).findAny().get()).isDisabled() ? R.string.menu_enable : R.string.menu_disable);
        }
        return true;
    }

    @Override
    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
        List<Model> selectedItems = new ArrayList<>(listManager.getSelectedItems());
        if (selectedItems.isEmpty()) {
            ACRA.getErrorReporter().putCustomData("listManager", listManager.toString());
            ACRA.getErrorReporter().handleSilentException(new IllegalStateException("No selected items"));
            return false;
        }
        switch (item.getItemId()) {
            case R.id.action_rename:
                ScriptUtils.renameDialog(scriptManager, getActivity(), listManager, selectedItems.get(0));
                break;
            case R.id.action_edit:
                ScriptUtils.editScript(getActivity(), listManager, (Script) selectedItems.get(0));
                break;
            case R.id.action_backup:
                Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
                intent.addCategory(Intent.CATEGORY_OPENABLE);
                intent.setType("application/javascript");
                Script script = (Script) selectedItems.get(0);
                intent.putExtra(Intent.EXTRA_TITLE, script.getId() + "_" + script.getName().replace("[,\\./\\:*?\"<>\\|]", "_") + ".js");
                startActivityForResult(intent, EXPORT);
                break;
            case R.id.action_format:
                ScriptUtils.format(scriptManager, getActivity(), listManager, selectedItems);
                break;
            case R.id.action_disable:
                ScriptUtils.toggleDisable(scriptManager, listManager, (Script) selectedItems.get(0));
                break;
        }
        return true;
    }

    @Override
    public void onDestroyActionMode(ActionMode mode) {
        listManager.deselectAll();
    }
}

File: app/src/main/java/com/faendir/lightning_launcher/multitool/scriptmanager/ScriptManagerFragment.kt
package com.faendir.lightning_launcher.multitool.scriptmanager

import android.app.Activity
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceManager
import android.view.*
import android.widget.FrameLayout
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.faendir.lightning_launcher.multitool.R
import com.faendir.lightning_launcher.multitool.event.ClickEvent
import com.faendir.lightning_launcher.multitool.fastadapter.Model
import org.acra.ACRA
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import java.util.*

/**
 * Created by Lukas on 22.08.2015.
 * Main activity of ScriptManager
 */
class ScriptManagerFragment : Fragment(), ActionMode.Callback {
    private lateinit var sharedPref: SharedPreferences
    private var enableMenu: Boolean = false
    private lateinit var layout: FrameLayout
    private lateinit var listManager: ListManager
    private var actionMode: ActionMode? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sharedPref = PreferenceManager.getDefaultSharedPreferences(activity)
        listManager = ListManager(activity!!) { this.setActionModeEnabled(it) }
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        layout = FrameLayout(activity!!)
        return layout
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
        loadFromLauncher()
    }

    override fun onStop() {
        EventBus.getDefault().unregister(this)
        super.onStop()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) = inflater.inflate(R.menu.menu_scriptmanager, menu)

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (!enableMenu) {
            Toast.makeText(activity, R.string.toast_menuDisabled, Toast.LENGTH_SHORT).show()
            return true
        }
        when (item!!.itemId) {
            R.id.action_restore -> {
                val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
                intent.addCategory(Intent.CATEGORY_OPENABLE)
                intent.type = "*/*"
                startActivityForResult(intent, IMPORT)
            }
            R.id.action_search -> ScriptUtils.searchDialog(activity!!, listManager)
            else -> return super.onOptionsItemSelected(item)
        }
        return true
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK) {
            when (requestCode) {
                IMPORT -> if (data != null && data.data != null) {
                    ScriptUtils.restoreFromFile(activity!!, listManager, data.data!!)
                }
                EXPORT -> if (data != null && data.data != null) {
                    ScriptUtils.backup(activity!!, listManager, listManager.selectedItems[0], data.data!!)
                }
            }
        }
    }

    private fun loadFromLauncher() {
        layout.removeAllViews()
        LayoutInflater.from(activity).inflate(R.layout.fragment_loading, layout)
        listManager.update {
            listManager.setAsContentOf(layout)
            enableMenu = true
        }
    }

    @Subscribe
    fun onReloadButton(event: ClickEvent) {
        if (event.id == R.id.button_reload) {
            sharedPref.edit().putInt(getString(R.string.pref_id), -1).apply()
            loadFromLauncher()
        }
    }

    private fun setActionModeEnabled(enable: Boolean) {
        if (enable) {
            if (actionMode == null) {
                actionMode = activity!!.startActionMode(this)
            } else {
                actionMode!!.invalidate()
            }
        } else if (actionMode != null) {
            actionMode!!.finish()
            actionMode = null
        }
    }

    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
        mode.menuInflater.inflate(R.menu.menu_context_scriptmanager, menu)
        onPrepareActionMode(mode, menu)
        return true
    }

    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
        val selectionCount = listManager.selectedItems.size
        if (selectionCount == 0) {
            mode.finish()
        }
        val isOneScript = selectionCount == 1
        menu.findItem(R.id.action_rename).isVisible = isOneScript
        menu.findItem(R.id.action_edit).isVisible = isOneScript
        menu.findItem(R.id.action_backup).isVisible = isOneScript
        menu.findItem(R.id.action_format).isVisible = true
        val disable = menu.findItem(R.id.action_disable).setVisible(isOneScript)
        if (isOneScript) {
            disable.setTitle(if (listManager.selectedItems[0].isDisabled) R.string.menu_enable else R.string.menu_disable)
        }
        return true
    }

    override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
        val selectedItems = ArrayList<Model>(listManager.selectedItems)
        if (selectedItems.isEmpty()) {
            ACRA.getErrorReporter().putCustomData("listManager", listManager.toString())
            ACRA.getErrorReporter().handleSilentException(IllegalStateException("No selected items"))
            return false
        }
        when (item.itemId) {
            R.id.action_rename -> ScriptUtils.renameDialog(activity!!, listManager, selectedItems[0] as Script)
            R.id.action_edit -> ScriptUtils.editScript(activity!!, listManager, selectedItems[0] as Script)
            R.id.action_backup -> {
                val intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
                intent.addCategory(Intent.CATEGORY_OPENABLE)
                intent.type = "application/javascript"
                val script = selectedItems[0] as Script
                intent.putExtra(Intent.EXTRA_TITLE, script.id.toString() + "_" + script.name.replace("[,\\./\\:*?\"<>\\|]", "_") + ".js")
                startActivityForResult(intent, EXPORT)
            }
            R.id.action_format -> ScriptUtils.format(activity!!, listManager, selectedItems)
            R.id.action_disable -> ScriptUtils.toggleDisable(listManager, selectedItems[0] as Script)
        }
        return true
    }

    override fun onDestroyActionMode(mode: ActionMode) = listManager.deselectAll()

    companion object {
        private const val IMPORT = 1
        private const val EXPORT = 2
    }
}
